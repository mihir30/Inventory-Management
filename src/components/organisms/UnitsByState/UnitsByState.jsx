/* eslint-disable array-callback-return */
import React, { useEffect, useState } from 'react'
import { getData } from '../../utils/common.utils'
import FilterSectionUnits from '../../molecules/FilterSectionUnits/FilterSectionUnits'
import NavigationBar from '../../molecules/NavigationBar/NavigationBar'
import UnitsByStateContentSection from '../../molecules/UnitsByStateContentSection/UnitsByStateContentSection'
import UnitsByStateEmptySection from '../../molecules/UnitsByStateEmptySection/UnitsByStateEmptySection'
import './UnitsByState.scss'
import '../../molecules/UnitsByStateContentSection/UnitsByStateContentSection.scss'

const UnitsByState = () => {

    const [unitsList, setUnitsList] = useState([]);
    const url = window.location.href;
    useEffect(() => {
        const fetchData = async () => {
            try {
                const units = await getData('http://localhost:8000/units');
                setUnitsList(units);

            } catch (error) {
                console.log("error", error.message);
            }
        };
        fetchData();

    }, [url]);

    const typeOfAllUnits = ["Monitor", "Computer", "Memory", "Hard Disk", "Processor"];
    const [fromDate, setItems] = useState([]);
    const [status, setStatus] = useState("Added");
    const [type,setType] = useState("All units")
    useEffect(() => {
        const fromDateList = JSON.parse(localStorage.getItem('dateList'));
        if (fromDateList) {
            setItems(fromDateList);
        }

        const statusvalue = JSON.parse(localStorage.getItem('status'));
        if (statusvalue) {
            setStatus(statusvalue);
        }

        const typevalue = JSON.parse(localStorage.getItem('type'));
        if (typevalue) {
            setType(typevalue);
        }
        

    }, [fromDate,status,type]);

    const filterUnitsSection = () => {

        const results = unitsList.filter(element => {
            if (type === "All units") {
                if (status === "Added") {
                    return fromDate.includes(element.fromDate) && (typeOfAllUnits.includes(element.type)) && (element.status === "Taken" || element.status === "Available");
                }
                else {
                    return fromDate.includes(element.fromDate) && (typeOfAllUnits.includes(element.type)) && element.status === status;
                }
            } else {
                if (status === "Added") {
                    return fromDate.includes(element.fromDate) && element.type === type && (element.status === "Taken" || element.status === "Available");
                }
                else {
                    return fromDate.includes(element.fromDate) && element.type === type && element.status === status;
                }
            }
        });

        return results;

    }

    const results = filterUnitsSection();



    return (
        <div><NavigationBar></NavigationBar>
            <FilterSectionUnits></FilterSectionUnits>
            <div className='units-by-state-section'>
                <strong className='units-text'>All units</strong>: {results.length}
                <div className='table-section'>
                    <div className="row">
                        <div className="table-responsive">
                            <table className="table table-fixed table-borderless table-hover bg-white">
                                <thead className="table-header-unitsbystate reports-table-thead">
                                    <tr className='reports-table-tr' >
                                        <th scope="col" className="col-2 status-column reports-table-th" style={{ paddingLeft: "3.2rem" }}>Status</th>
                                        <th scope="col" className="col-2 serial-column reports-table-th ">Serial</th>
                                        <th scope="col" className="col-2 type-column reports-table-th ">Type</th>
                                        <th scope="col" className="col-2 name-column reports-table-th ">Name</th>
                                        <th scope="col" className="col-2 assigned-column reports-table-th ">Assigned to</th>
                                        <th scope="col" className="col-2 reports-table-th ">From</th>
                                    </tr>
                                </thead>
                                <tbody className='reports-table-tbody'>
                                    {results && results.length > 0 ?
                                        results.map((unit) => (
                                            <UnitsByStateContentSection key={unit.id}
                                                id={unit.id}
                                                status={unit.status}
                                                serial={unit.serial}
                                                type={unit.type}
                                                name={unit.name}
                                                person={unit.person}
                                                fromDate={unit.fromDate}></UnitsByStateContentSection>
                                        ))
                                        :
                                        <UnitsByStateEmptySection></UnitsByStateEmptySection>
                                    }
                                </tbody>
                            </table>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    )
}

export default UnitsByState