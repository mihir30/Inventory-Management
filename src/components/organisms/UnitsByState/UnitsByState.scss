.units-by-state-section{
    margin:10px 0px;
    width: 1150px;
    font-size: 14px;
}

.table-section{
    position: relative;
    border-width: 0px;
    margin-left: 4px;
    margin-top: 3px;
    width: 1150px;
    height: 425px;
    z-index: 0;
    overflow: hidden;
    background-color: white;
}

.units-text{
    margin: 10px 0px 10px 30px;
}

.reports-table-tbody {
    height: 545px;
    overflow-y: auto;
    width: 100%;
    
}

.reports-table-thead,
.reports-table-tbody,
.reports-table-tr,
.reports-table-td,
.reports-table-th {
    display: flex;font-size:14px;
    background-color: white;
}
.wrench-icon {
    margin-right: 1rem;
    color: grey;
}
.reports-table-tbody,
.reports-table-td {
    display: inline-block;
    background-color: white;
    
}

.reports-table-tbody,
.reports-table-th,
.reports-table-thead > tr > th {
    float: left;
    position: relative;
    background-color: #f4f5f9;
}

.table-header-unitsbystate {
    height: 43px;
    background: #f4f5f9;
    color: #313131;
    font-weight: 500;
    letter-spacing: .2px;
    clear: both;
    overflow: hidden;
    border-bottom: 1px solid #5ccce7;
    width:100%;
}



.status-column{
    width:410px;
}

.serial-column{
    width: 185px;
}

.type-column{
    width: 185px;
}

.name-column{
    width: 185px;
}

.assigned-column{
    width: 185px;
}

