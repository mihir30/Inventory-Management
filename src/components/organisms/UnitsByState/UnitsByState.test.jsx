import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import ReactDom from "react-dom";
import UnitsByState from './UnitsByState';

describe("Units by state Component", () => {
    const typeOfAllUnits = ["Monitor", "Computer", "Memory", "Hard Disk", "Processor"];
    const type="All units";
    const status="Added";
    const fromDate="2020-01-02";

    const filterUnitsSection = jest.fn(() => {
        const results = [{type:"Monitor", status:"Available", fromDate:"2020-01-02"},{type:"Computer", status:"Available", fromDate:"2020-01-03"}].filter(element => {
            if (type === "All units") {
                if (status === "Added") {
                    return fromDate.includes(element.fromDate) && (typeOfAllUnits.includes(element.type)) && (element.status === "Taken" || element.status === "Available");
                }
                else {
                    return fromDate.includes(element.fromDate) && (typeOfAllUnits.includes(element.type)) && element.status === status;
                }
            } else {
                if (status === "Added") {
                    return fromDate.includes(element.fromDate) && element.type === type && (element.status === "Taken" || element.status === "Available");
                }
                else {
                    return fromDate.includes(element.fromDate) && element.type === type && element.status === status;
                }
            }
        });

        return results
    })

    const results = filterUnitsSection();

    it('renders the component properly', () => {
        const divEl = document.createElement("div");
        ReactDom.render(<BrowserRouter><UnitsByState></UnitsByState></BrowserRouter>, divEl); 
    });

    it('should return results correctly', () => {
        expect(results).toEqual([{type:"Monitor", status:"Available", fromDate:"2020-01-02"}]);
    })

    it('should have the right units by state component snapshot', () => {
        const tree = renderer.create().toJSON();
        expect(tree).toMatchSnapshot();
    })

});