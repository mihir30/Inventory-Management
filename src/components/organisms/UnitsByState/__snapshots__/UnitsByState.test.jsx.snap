// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Units by state Component should have the right units by state component snapshot 1`] = `null`;
